src/docker-extract.c

```diff
@@ -0,0 +1,320 @@
+    archive_read_support_compression_all(a);
+#else
+    archive_read_support_filter_all(a);
+#endif
+    archive_read_support_format_all(a);
+    ret = archive_read_open_filename(a, tarfile, 10240);
+    if (ret != ARCHIVE_OK)
+        return (1);
+
+    while (archive_read_next_header(a, &entry) == ARCHIVE_OK) {
+
+        const char *pathname = archive_entry_pathname(entry);
+
+        if (strstr(pathname, "/.wh..wh..opq")) {
+            singularity_message(DEBUG, "Opaque Marker %s\n", pathname);
+            errcode = apply_opaque(pathname, rootfs_dir);
+            if (errcode != 0) {
+                break;
+            }
+        } else if (strstr(pathname, "/.wh.")) {
+            singularity_message(DEBUG, "Whiteout Marker %s\n", pathname);
+            errcode = apply_whiteout(pathname, rootfs_dir);
+            if (errcode != 0) {
+                break;
+            }
+        }
+    }
+#if ARCHIVE_VERSION_NUMBER <= 3000000
+    ret = archive_read_finish(a);
+#else
+    ret = archive_read_free(a);
+#endif
+    if (ret != ARCHIVE_OK){
+        singularity_message(ERROR, "Error freeing archive\n");
+        ABORT(255);
+    }
+
+    return errcode;
+}
+
+/* See  https://github.com/libarchive/libarchive/wiki/Examples#A_Complete_Extractor */
+static int copy_data(struct archive *ar, struct archive *aw) {
+    int r;
+    const void *buff;
+    size_t size;
+    int64_t offset;
+
+    for (;;) {
+        r = archive_read_data_block(ar, &buff, &size, &offset);
+        if (r == ARCHIVE_EOF)
+            return (ARCHIVE_OK);
+        if (r < ARCHIVE_OK)
+            return (r);
+        r = archive_write_data_block(aw, buff, size, offset);
+        if (r < ARCHIVE_OK) {
+            singularity_message(ERROR, "tar extraction error: %s\n", archive_error_string(aw));
+            return (r);
+        }
+    }
+}
+
+/* extract_tar
+ *  Extract a tar file to rootfs_dir using libarchive. Handles compression.
+ *  Exclude any .wh. whiteout files, and device/pipe/fifo entries
+ *
+ * See https://github.com/libarchive/libarchive/wiki/Examples#A_Complete_Extractor
+ */
+int extract_tar(const char *tarfile, const char *rootfs_dir) {
+    int retval = 0;
+    struct archive *a;
+    struct archive *ext;
+    struct archive_entry *entry;
+    int flags;
+    int r;
+    char *orig_dir;
+
+    orig_dir = get_current_dir_name();
+
+    /* Select which attributes we want to restore. */
+    flags = ARCHIVE_EXTRACT_TIME;
+    flags |= ARCHIVE_EXTRACT_PERM;
+    flags |= ARCHIVE_EXTRACT_ACL;
+    flags |= ARCHIVE_EXTRACT_FFLAGS;
+
+    a = archive_read_new();
+    archive_read_support_format_all(a);
+#if ARCHIVE_VERSION_NUMBER <= 3000000
+    archive_read_support_compression_all(a);
+#else
+    archive_read_support_filter_all(a);
+#endif
+    ext = archive_write_disk_new();
+    archive_write_disk_set_options(ext, flags);
+    archive_write_disk_set_standard_lookup(ext);
+    if ((r = archive_read_open_filename(a, tarfile, 10240))){
```

Comment:
AFAIK the value here doesn't really matter, except for performance reasons.  In the IO callback docs they specify:

>To read an archive, you need to provide libarchive with a _read callback_ function. The read callback is expected to provide a _block_ of data to libarchive each time it is called.

>Blocks can be of any size or alignment at all. There are only two requirements:

>Each block must be at least one byte long. (A zero-byte block indicates end-of-file.)
When you return a pointer to a block, the memory where the block is stored must be stable until the next time your read function is called.

https://github.com/libarchive/libarchive/wiki/LibarchiveIO

From there it goes on to mention performance tuning, matching hardware block size etc. but there is no hard requirement on the block size.

For libarchive 3+ it may not use the specified value - but will inspect the archive for a block size to use:

>Note 1: Beginning with libarchive 3.0, the `archive_read_open_filename()` function inspects the file before deciding how to handle the blocksize. If the filename provided refers to a tape device, for example, it will use exactly the blocksize you specify. For other devices, it may adjust the requested blocksize in order to obtain better performance.

https://github.com/libarchive/libarchive/wiki/Examples

The 10240 is referenced elsewhere in docs as default for *write* for compatibility with old tar versions, and this is taken from the 'universal reader' example.

Reply:
Makes sense.  Thanks for the clarification and the pointer to the docs.  😸